Enable SVG Format Graph Export

capricornunicorn123 opened the pull request:
![Final image](https://github.com/constellation-app/constellation/assets/139201011/6746ce07-d2b6-4bb2-ab9b-d707b78a3d71)

<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [x] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->
This PR facilitates the export of graph data to an SVG format to enable information to be visualized outside of constellation whilst optimizing resolution.


### Alternate Designs

<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->
The PR has been developed in 8 key Milestones as detailed below:
- [x] [M1] Enable Nodes to be exported to a SVG.
- [x] [M2] Enable Transactions to be exported to a SVG.
- [x] [M3] Enhance Nodes to include node visual attributes.
- [x] [M4] Enhance Transactions to include transaction visual attributes including edge link and transaction views.
- [x] [M5] Enable export to factor in z-axis scaling and correct element overlaying
- [x] [M6] Enable export to factor in x and y repositioning based on z-axis scaling.
- [x] [M7] Enable export filtering.
- [x] [M8] Enable Export to factor in camera perspective.

Design discussions can be seen during the review process for each of these milestones. 

### Why Should This Be In Core?

<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->
Exporting graphs is a core feature of Constellation.

### Benefits

<!-- What benefits will be realized by the code change? -->

This enhancement stands to replace the export to screenshot be enriching the quality of exported images. Being able to export large amounts of data in a universally viewable format and enable anyone to view the data from a common browser.

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

These exported files increase linearly with graph complexity. The export interface has been designed to intuit user interaction however some users may find that the functionality may need some explication.

### Verification Process

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->
During the development of the feature, a number of developers were invited to review and contribute to the functionality of the feature. They were able to bring their individual experience with the application and its users to ensure that the feature covers as many use cases as possible. 

### Applicable Issues
#1119 
<!-- Link any applicable issues here -->

capricornunicorn123 commented:
### Milestone 1
![Milestone 1 Graphic](https://github.com/constellation-app/constellation/assets/139201011/c3d435b2-2d8b-489d-8ab9-d23607295f00)

**Summary:**
PR enables an active graph to be interpreted and exported as an SVG file that can be rendered in common browsers. 
Translation is limited to Node elements positioned on an x, y plane. Node scale and position is accurately translated between graph and SVG.

**limitations & Potential Issues:**
- Exporting 3d graphs produces undesirable results due to lack of consideration for z axis. 
- Non positional attributes such as color and ID are currently excluded from the output.
- Transactions are currently excluded from the output.
- Code smells relating to OS supportability have been highlighted in the current use of FileReader and InputStreamReader.
- Overwriting a file on export appears to be buggy only working in some instances.
- SVG parser only permits loading SVG files that utilize single line tags.
- The Plugin has a lot of responsibility regarding SVG generation and is currently inefficient on larger graphs due to the repetitive iteration. 
- Template files are accessed by building a file path from an existing class located in the same package as the same file path.

**Possible Improvements:**
- Transition the current responsibility of node generation and extraction from the plugin to separate helper class (perhaps SVGBuilder). This would permit some persistent variable such as graph boundaries and template file locations to be accessed one for the life of the plugin execution instead of multiple times.
- Relocate resource SVG files to a separate resource package.

*Note: Tests have not yet been implemented due to the likelihood for major code refactoring over the coming days.*

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGAction.java:
```diff
@@ -0,0 +1,94 @@
+    }
+   
+    @Override
+    public void actionPerformed(ActionEvent e) {
```
parameter final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,190 @@
+    }
+    
+    @Override
+    protected void read(GraphReadMethods graph, PluginInteraction interaction, PluginParameters parameters) throws InterruptedException, PluginException { 
```
parameters final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,190 @@
+            
+            final int vertexID = graph.getVertex(vertexPosition);
+            
+            Float xVal = (graph.getFloatValue(xAttributeID, vertexID) * 128) - bounds[0][0];
+            Float yVal = (bounds[1][1] - bounds[1][0]) - ((graph.getFloatValue(yAttributeID, vertexID) * 128) - bounds[1][0]);
```
finals

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,190 @@
+        ArrayList<SVGObject> nodes = new ArrayList<>();
+        Float[][] bounds = getBounds(graph);
+        
+        int xAttributeID = VisualConcept.VertexAttribute.X.get(graph);
+        int yAttributeID = VisualConcept.VertexAttribute.Y.get(graph);
+
+        int vertexCount = graph.getVertexCount();
```
finals

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,190 @@
+        SVGObject node = null;
+        try {
+            node = SVGParser.parse(inputStream);
+        } catch (IOException ex) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,190 @@
+     * @return 
+     */
+    private SVGObject generateNode(final SVGObject parent) { 
+        SVGResourceConstants resourceClass = new SVGResourceConstants();
+        InputStream inputStream = resourceClass.getClass().getResourceAsStream(SVGResourceConstants.NODE);
```
finals

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,202 @@
+    private String elementToSVG(final String prefix){
+        final StringBuilder attributeBuilder = new StringBuilder();
+        final String linePrefix = SeparatorConstants.NEWLINE + prefix;
+        Set<String> keys = attributes.keySet();
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,202 @@
+            linePrefix += prefix;
+        }
+        
+        StringBuilder attributeBuilder = new StringBuilder();
+        
+        Set<String> keys = attributes.keySet();
```
finals

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGParser.java:
```diff
@@ -0,0 +1,170 @@
+        SVGObject currentElement = null; 
+        final Collection<SVGObject> roots = new HashSet<>();
+        
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream))) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGParser.java:
```diff
@@ -0,0 +1,170 @@
+            String line;
+            while ((line = br.readLine()) != null) {
+                final String svgElement = SVGParser.isolateSVGElement(line);
+                boolean openTag = SVGParser.isOpenTag(svgElement);
+                boolean closeTag = SVGParser.isCloseTag(svgElement);
```
finals

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGParser.java:
```diff
@@ -0,0 +1,170 @@
+   
+    private static Map<String,String> getElementAttributes(final String svgString) {
+        Map<String,String> attributes = new HashMap<>();
+        String stripped = stripAngleBrackets(svgString);
+        String[] components = stripped.split(SeparatorConstants.BLANKSPACE);
+        for (String component : components) {
+            if (component.contains("=")) {
+                String[] attribute = component.split("=");
+                if (attribute.length == 2){
+                    String key = attribute[0];
+                    String value = attribute[1].replaceAll(SeparatorConstants.QUOTE, "");
```
I think all of these strings can be final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGParser.java:
```diff
@@ -0,0 +1,170 @@
+     * @return 
+     */
+    private static String isolateSVGElement(final String line) {
+        String svgElement = line.substring(line.indexOf("<"), line.indexOf(">") + 1);
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGParser.java:
```diff
@@ -0,0 +1,170 @@
+     * @return 
+     */
+    private static String stripAngleBrackets(final String line) {
+        String svgElement = line.substring(line.indexOf("<") + 1, line.indexOf(">"));
```
final

Delphinus8821 commented:
Looking really good so far. The basic export plugin works and the code is set out nicely, easy to read and well commented.

capricornunicorn123 commented:
### Milestone 2
 
![Milestone 2 Graphic](https://github.com/constellation-app/constellation/assets/139201011/74cf9e98-a9e6-4ab6-a58c-b8680148ffc6)

**Summary:**
Output SVG generation has been improved to include links, increase efficiency, encapsulate component responsibility and to enable a closer resemblance to the active graph. Translation from graph to SVG has been extended to links positioned on an x, y plane.  

**Improvements Made:**
- Overwriting a file on export now works and prompts the user for approval to do so.  This option does not exist for any other export or save options and may be useful to incorporate to prevent users accidentally overwriting files.
- Responsibility of SVG generation has been abstracted away from the `ExportToSVGPlugin` and as a result a `SVGGraph` and corresponding `SVGGraphBuilder` has been created. 
- SVG files have been encapsulated in a resource package along with a `SVGResourceConstant` enum class. This permits constants representing a file name to be verified as the `SVGResourceConstant` type and for the class path to be built using this `SVGResourceConstant`.
- SVG output files are now build around a `Layout.svg` to encapsulate renderbale areas of the output image. Of note, the output file has SVG components of class type header, content and footer.

**limitations & Potential Issues:**
- *Exporting 3D graphs produces undesirable results due to lack of consideration for z axis.*
- *Non positional attributes such as color and ID are currently excluded from the output.*
- *Code smells relating to OS supportability have been highlighted in the current use of `FileReader` and `InputStreamReader`.*
- *`SVGParser` only permits loading SVG files that utilize single line tags.*
- *`SVGParser` will not accept template SVG files that contain only one line.*
- Transactions are rendered to the screen as a single line to verify that the correct dimensions have been identified, rendering all of the different possible transaction directions and combination groups dynamically will be a challenge.
- Element attributes are accessed using string literals that are “hard coded” into each instance they are used. 
- SVG child elements are currently stored in a data structure of `HashMap<String, ArrayList<SVGObject>>`, where the key of the HashMap is the `SVGObject`'s class attribute value. This is complex and may not serve any additional benefit over storing elements by their ID which should be unique and by extension does not require such a complex data structure.
- `SVGContainer`'s are not currently able to set their size based on the content contained within them. 

**Possible Improvements:**.
- Create some form of translator module that is able to return the SVG equivalent attribute key for a Constellation graph attribute name and vice versa. 
- Refactor `SVGObjects` to be sorted and accessed by ID and not class.
- Incorporate the `PluginReporter` into the plugin.
- The responsibility of setting the dimensions of a `SVGContainer` would be better served in the `SVGContainer` wrapper rather than the `SVGGraphBuilder`. This will become more apparent when graph titles and node titles are implemented.

*Note: Tests have not yet been implemented due to the likelihood for major code refactoring over the coming days.*

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGAction.java:
```diff
@@ -0,0 +1,133 @@
+                })
+                .setSelectionApprover(new SelectionApprover() {
+                    @Override
+                    public boolean approve(File[] files) {
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGAction.java:
```diff
@@ -0,0 +1,133 @@
+                    @Override
+                    public boolean approve(File[] files) {
+
+                        for (File file : files){
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,85 @@
+    protected void read(final GraphReadMethods graph, final PluginInteraction interaction, final PluginParameters parameters) throws InterruptedException, PluginException { 
+        final String fnam = parameters.getStringValue(FILE_NAME_PARAMETER_ID);
+        final File imageFile = new File(fnam);     
+        SVGObject svg = new SVGGraph.SVGGraphBuilder().withGraph(graph).build();
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,85 @@
+        SVGObject svg = new SVGGraph.SVGGraphBuilder().withGraph(graph).build();
+        try {
+            exportToSVG(imageFile, svg);
+        } catch (IOException ex) {
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,285 @@
+            SVGObject templateSVG = null;
+            try {
+                templateSVG = SVGParser.parse(inputStream);
+            } catch (IOException ex) {
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,285 @@
+            try {
+                templateSVG = SVGParser.parse(inputStream);
+            } catch (IOException ex) {
+                Exceptions.printStackTrace(ex);
```
Replace this with either a LOGGER call or a user Notification call.

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,226 @@
+    private final Map<String, ArrayList<SVGObject>> children;
+    private SVGObject parent;
+
+    public SVGObject(String type, SVGObject parent, Map<String, String> attributes) {
```
3x `final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,226 @@
+        if (prefix != null){
+            childPrefix += prefix;
+        }
+        for (SVGObject child : this.getAllChildren()){
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGAttributeConstant.java:
```diff
@@ -0,0 +1,50 @@
+    
+    private final String key;
+    
+    private SVGAttributeConstant(String attributeKey){
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGFileNameConstant.java:
```diff
@@ -0,0 +1,35 @@
+    
+    public final String resourceName;
+        
+    private SVGFileNameConstant(String resourceName){
```
`final`

OrionsGuardian commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGLayoutConstant.java:
```diff
@@ -0,0 +1,39 @@
+    
+    public final String name;
+        
+    private SVGLayoutConstant(String name){
```
`final`

capricornunicorn123 commented:
### Milestone 3
 
![Milestone 3 Graphic](https://github.com/constellation-app/constellation/assets/139201011/74fd907c-5e18-41eb-b1b6-45a882eb9d7b)

**Summary:**
Enhanced node output to include all Image and labeling elements.

**Improvements Made:**
- Nodes now export with background image, foreground image and decorator images.
- Nodes now contain top and bottom labels that size and color acurately.
- Warning dialog prevents users from exporting an empty graph.
- Exported SVG contains a header and footer element with title, subtitle and disclaimer.
- Elements are now rendered in the order that they are generated during the `build( )` process.

**limitations & Potential Issues:**
- *Exporting 3D graphs produces undesirable results due to lack of consideration for z axis.*
- *Non positional attributes such as color and ID are currently excluded from the output.*
- *Code smells relating to OS supportability have been highlighted in the current use of `FileReader` and `InputStreamReader`.*
- *`SVGParser` only permits loading SVG files that utilize single line tags.*
- *`SVGParser` will not accept template SVG files that contain only one line.*
- Transactions are rendered to the screen as a single line to verify that the correct dimensions have been identified, rendering all of the different possible transaction directions and combination groups dynamically will be a challenge.
- Element attributes are accessed using string literals that are “hard coded” into each instance they are used. 
- SVG child elements are currently stored in a data structure of `HashMap<String, ArrayList<SVGObject>>`, where the key of the HashMap is the `SVGObject`'s class attribute value. This is complex and may not serve any additional benefit over storing elements by their ID which should be unique and by extension does not require such a complex data structure.
- `SVGContainer`'s are not currently able to set their size based on the content contained within them. 
- - special characters
- padding for overflow not dynamic

**Possible Improvements:**.
- Create some form of translator module that is able to return the SVG equivalent attribute key for a Constellation graph attribute name and vice versa. 
- Refactor `SVGObjects` to be sorted and accessed by ID and not class.
- Incorporate the `PluginReporter` into the plugin.
- The responsibility of setting the dimensions of a `SVGContainer` would be better served in the `SVGContainer` wrapper rather than the `SVGGraphBuilder`. This will become more apparent when graph titles and node titles are implemented.

*Note: Tests have still not been implemented due to time constraints. *

altair1673 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,207 @@
+                if (!openTag && !closeTag) {
+                    throw new UnsupportedOperationException(String.format("This line could not be interpreted: %s", svgElement));
+                }
+                
```
Possible bug here. If the layout.svg file is parsed, this code attempts to add the first SVG element (the one with the xmlns attribute) twice. Although it won't be added in twice because the Set is used to store the elements maybe refactor it so the second attempt is not made. This would reduce the chance of bugs and make the code clearer for anyone in who works on it in the future.

altair1673 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,481 @@
+         * @param graph
+         * @return 
+         */
+        private void defineBoundary(final GraphReadMethods graph) {
```
Move this functionality to buildNodes() because you are looping through all nodes here a second time. Just calculate the bounds when you are initially looping through the nodes to get all the other attributes.

capricornunicorn123 replied:
The boundary needs to be defined before nodes are created as the SVG coordinates need to be adjusted from the Graph coordinates with respect to the boundary. The boundary needs to be known before this step. To simplify this method the `BoundingBox` and associated utilities have been used to calculate the extremities of the graph. If efficiency is an issue this bounding box could be improved to persist for the life of all graphs and be included in GraphReadMethods, it could then be updated dynamically as the graph is altered. Current implementations appear to generate a new `BoundingBox` as needed.

altair1673 reviewed (commented):
Looks good so far. Probably wouldn't do the whole parsing in existing file and adding stuff to it but maybe in this case it easier for some reason and I just cannot see it.

capricornunicorn123 commented:
### Milestone 4
![Milestone 4 Graphic](https://github.com/constellation-app/constellation/assets/139201011/35210c1b-d959-4a6a-ad14-76c3b0cc99d8)

**Summary:**
Enhanced connection output to include accurate representations of Transactions, Edges and Links.

**Improvements Made:**

- SVG template file parsing now uses regex’s.
- SVG template files can now include blank lines with no impact of interpreted SVG data.
- Links, Connections and Transactions generate with correct color, size, position and arrowhead. 
- SVG export factors in current connection mode.
- All graph positional references are stored in Tuples instead of individual coordinate variables.
- Setting SVG attributes has been simplified with a number of overload methods allowing attributes to store raw datatypes `int`, `Double` and `Float` as a `String` without string formatting.
- `BoundingBox` and associated helper class are used to determine the occupied area of Constellation graph visual content.
- `SVGObject` children are now stored by id in a `LinkedHashMap<String, SVGobject>` to preserve their creation order and enforce uniqueness. 

**limitations & Potential Issues:**

- Exporting 3D graphs produces undesirable results due to lack of consideration for z axis.
- Code smells relating to OS supportability have been highlighted in the current use of FileReader and InputStreamReader.
- SVGParser only permits loading SVG files that utilize single line tags.
- SVGParser will not accept template SVG files that contain only one line.
- SVGContainer's are not currently able to set their size based on the content contained within them.
- The current SVG output is not able to render non Latin characters.
- Padding has been set manually and is static, should be dynamic based on content.
- Custom background images are overwritten by the color value of the background icon for the respective node.
- Setting a decorator to an attribute with a value that doesn't correlate to an image results in a "?" icon and should show nothing instead.
- Transaction labels have not been implemented.
- Transaction counts for links and edges have not been implemented.
- Transaction arrow shaft does not support custom line styles.
- Looped transactions not supported.
- Some child `SVGObjects` have id null.
- Transaction Dimming not implemented. 
- Node Dimming not implemented.
- Connection generation produces undesirable results when nodes are placed in close proximity to one another.

**Possible Improvements:**
- Create some form of translator module that is able to return the SVG equivalent attribute key for a Constellation graph attribute name and vice versa.
- Incorporate the PluginReporter into the plugin.
- The responsibility of setting the dimensions of a SVGContainer would be better served in the SVGContainer wrapper rather than the SVGGraphBuilder.
- Instead of having 4 different methods called by the buildConnections(0method, group similar functionality and reduce the number of methods.
- The SVGGraphBuilder is bloating in size, abstracting the build methods into more relevent classes may be useful. i.e. buldNodes logi could be abstracted into a NodeBuilder in an SVGNode class that is tightly coupled to the Node.svg template file and manages its own building logic…

*Note: Tests have still not been implemented due to time constraints.*
**Reviewer: **

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGAction.java:
```diff
@@ -0,0 +1,140 @@
+                    fnam += FileExtensionConstants.SVG;
+                }
+
+                int connectionModeID = VisualConcept.GraphAttribute.CONNECTION_MODE.get(graph);
+                ConnectionMode currentConnectionmode = graph.getObjectValue(connectionModeID, 0);
```
finals

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGAction.java:
```diff
@@ -0,0 +1,140 @@
+                
+                PluginExecution.withPlugin(ImportExportPluginRegistry.EXPORT_SVG)
+                        .withParameter(ExportToSVGPlugin.FILE_NAME_PARAMETER_ID, fnam)
+                        .withParameter(ExportToSVGPlugin.GRAPH_TITLE_PARAMETER_ID, "Milestone 4")
```
I'm just assuming this will be updated by the end but just bringing to your attention if didn't already know

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,110 @@
+@PluginInfo(pluginType = PluginType.EXPORT, tags = {PluginTags.EXPORT})
+@NbBundle.Messages("ExportToSVGPlugin=Export to SVG")
+public class ExportToSVGPlugin extends SimpleReadPlugin {
+    public static final String FILE_NAME_PARAMETER_ID = PluginParameter.buildId(ExportToSVGPlugin.class, "filename");
+    public static final String GRAPH_TITLE_PARAMETER_ID = PluginParameter.buildId(ExportToSVGPlugin.class, "graphtitle");
+    public static final String GRAPH_CONNECTION_MODE_PARAMETER_ID = PluginParameter.buildId(ExportToSVGPlugin.class, "connectionmode");
```
are you able to use underscore notation on the parameter ids to keep consistent with the naming convention used elsewhere? i.e. `file_name`, `graph_title`, `connection_mode`

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,110 @@
+        final PluginParameters parameters = new PluginParameters();
+
+        final PluginParameter<FileParameterValue> fnamParam = FileParameterType.build(FILE_NAME_PARAMETER_ID);
+        fnamParam.setName(FILE_NAME_PARAMETER_ID);
```
I would set the string to "File Name" as the string built for the parameter won't be what you think it is. Similar case for the other 2 parameters

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/ExportToSVGPlugin.java:
```diff
@@ -0,0 +1,110 @@
+        try {
+            exportToSVG(imageFile, svg);
+        } catch (final IOException ex) {
+            Exceptions.printStackTrace(ex);
```
could you replace this with a logger entry?

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGContainer.java:
```diff
@@ -0,0 +1,125 @@
+     * Sets the height of the container element.
+     * @param height
+     */
+    private void setHeight(final Float height) {
+        svgObjectReference.setAttribute(SVGAttributeConstant.HEIGHT.getKey(), height.toString());
+    }
```
if you wanted to use the primitive `float` here, you could use `String.valueOf(height)`. If null is valid though, note that `height.toString()` throws a null pointer if `height` is null so you'll need to add a null check there. Will be a similar case for the other similar functions

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGContainer.java:
```diff
@@ -0,0 +1,125 @@
+     * Sets the width of the container element.
+     * @param width 
+     */
+    private void setWidh(final Float width) {
```
setWidth

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGContainer.java:
```diff
@@ -0,0 +1,125 @@
+     * @param x
+     * @param y 
+     */
+    public void setposition(final Float x, final Float y) {
```
setPosition

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+         * Other graph state factors including maxTransactions and drawFlags are considered.
+         * @param svgGraph 
+         */
+        private void buildConnections(SVGGraph svgGraph) {
```
parameter final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+                        int directedEdgeCount = 0;
+                        final int linkEdgeCount = graph.getLinkEdgeCount(linkID);
+                        for (int position = 0; position < linkEdgeCount; position++){
+                            int edgeID = graph.getLinkEdge(linkID, position);
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+         * @param connectionsContainer
+         * @param linkID
+         */
+        private void buildSingleLink(final SVGContainer connectionsContainer, int linkID) {
```
`linkID` can be final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+         * @param connectionsContainer
+         * @param linkID
+         */
+        private void buildDoubleLink(final SVGContainer connectionsContainer, int linkID) {
```
same here

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+                }
+                
+                //Determine offset controlls
+                int paralellOffsetDistance = (processedEdgeCount / 2 + ((processedEdgeCount % 2 == 0) ? 0 : 1)) * 16;
+                double paralellOffsetDirection = Math.pow(-1, processedEdgeCount);
+                double paralellOffsetAngle = Math.toRadians(90) * paralellOffsetDirection;
```
finals

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+         * @param connectionsContainer
+         * @param linkID
+         */
+        private void buildTransactions(final SVGContainer connectionsContainer, int linkID) {
```
`linkID` final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+            final int transactionCount = graph.getLinkTransactionCount(linkID);
+            for (int transactionPosition = 0 ; transactionPosition < transactionCount; transactionPosition++) {
+                final int transactionID = graph.getLinkTransaction(linkID, transactionPosition); 
+                int transactionDirection = graph.getTransactionDirection(transactionID);
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+         * @param y
+         * @param connectionAngle 
+         */
+        private void buildArrowHead(SVGObject arrowHeadContainer, Tuple<Double,Double> position, Double connectionAngle) {
```
parameters final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+            final GraphLabels bottomLabelsContainer = graph.getObjectValue(bottomLabelsAttributeID, 0);
+            final List<GraphLabel> labels = bottomLabelsContainer.getLabels();
+            
+            Integer offset = 0;
```
could probably use primitive `int` here

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+            final GraphLabels topLabelsContainer = graph.getObjectValue(topLabelsAttributeID, 0);
+            final List<GraphLabel> labels = topLabelsContainer.getLabels();
+            
+            Integer offset = 0;
```
same here

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+                if (labelString != null){
+                    final SVGObject text = buildSVGObjectFromTemplate(SVGFileNameConstant.TOP_LABEL);
+                    final Float size = label.getSize() * 64;
+                    final Integer intSize = size.intValue();
```
same here

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+            try {
+                templateSVG = SVGParser.parse(inputStream);
+            } catch (final IOException ex) {
+                LOGGER.log(Level.INFO, Arrays.toString(ex.getStackTrace()));
```
use `ex.getLocalisedMessage()` instead

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,894 @@
+                if (attributeID != Graph.NOT_FOUND){
+                    final String attributeValue;
+                    if ("pinned".equals(vertexAttributeReference)){
+                        if ("true".equals(graph.getStringValue(attributeID, vertexID))){
```
if the attribute is pinned, you should be able to use `graph.getBooleanValue`

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,286 @@
+     * @return ArrayList
+     */
+    public List<SVGObject> getAllChildren() {
+        final ArrayList<SVGObject> allChildren = new ArrayList<>();
```
can you make the declaration a plain old list? i.e. `List<SVGObject>`? Instantiation is fine

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,217 @@
+     * @return
+     * @throws IOException 
+     */
+    public static SVGObject parse(InputStream inputStream) throws IOException {
```
parameter final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,217 @@
+                } 
+                // This parser curently requires all lines with be an SVG tag as it does not support multi line tags.
+                if (!openTag && !closeTag) {
+                    throw new UnsupportedOperationException(String.format("This line could not be interpreted: %s", svgElement));
```
I think the exception you want to be throwing here is `IllegalStateException`

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,217 @@
+            }
+        }
+        if (roots.size() != 1) {
+            throw new UnsupportedOperationException(String.format("The SVG file has %s outer elements.", roots.size()));
```
not sure exactly what exception to throw here but generally speaking I normally associate `UnsupportedOperationException` being applied to an entire function implementation rather than a portion. Perhaps `IllegalStateException` is better suited here?

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,217 @@
+     * @return 
+     */
+    private static String removeNonLatinCharacters(final String text){
+        StringBuilder builder = new StringBuilder();
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,217 @@
+     * @return 
+     */
+    private static String isolateSVGElement(final String line) {
+        final ArrayList<String> svgElements = new ArrayList<>();
```
declare as list

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/parser/SVGParser.java:
```diff
@@ -0,0 +1,217 @@
+            foundElements++;
+        }
+        if (foundElements > 1){
+            throw new UnsupportedOperationException("Found multiple SVG Elements");
```
similar case with exception

antares1470 reviewed (changes requested):
Looking good noting where you're currently at. The only other thing I would add is that I would look at some of the uses of `Float` and `Double` and other object version of primitives and consider whether the primitive can be used instead. Basically just have to consider whether null is a valid value to consider and if not, use the primitive type (as that one will generally be more performant)

capricornunicorn123 commented:
### Milestone 7
![Milestone 7 Graphic](https://github.com/constellation-app/constellation/assets/139201011/f77b3163-330e-465c-8f8e-86c96e49c75b)
**Summary:**
Enables customised graph exporting with plugin parameters.

**Improvements Made:**
- Refactored graph calculations and referencing to use GraphVisualAccess and BoundingBox utility classes.
- Enables user to customize graph parameters including, background colors, graph titles and element inclusion during export.
- Refactored SVGExport framework to better encapsulate responsibility.
    - SVGData: a minimalistic data structure containing foundational references to SVG element attributes, content, and relationships. 
    - SVGObject: A wrapper class for SVGData to enable simplified use when programing providing a number of helper methods.
    - SVGGraph: controls the generation of a SVG representation of a Constellation Graph
    - Connection generation is not encapsulated in a two functions.
- Single line SVG template files are now supported.
- Node Dimming enabled.
- Invisible Nodes no longer export.
- Connection line styles of dashed and dotted are now supported.

**limitations & Potential Issues:**
- *Exporting 3D graphs produces undesirable results due to lack of consideration for z axis.
- *Code smells relating to OS supportability have been highlighted in the current use of FileReader and InputStreamReader.*
- *SVGParser only permits loading SVG files that utilize single line tags.*
- *SVGContainer's are not currently able to set their size based on the content contained within them.*
- *The current SVG output is not able to render non Latin characters.*
- *Padding has been set manually and is static, should be dynamic based on content.*
- *Custom background images are overwritten by the color value of the background icon for the respective node.*
- *Setting a decorator to an attribute with a value that doesn't correlate to an image results in a "?" icon and should show nothing instead.*
- *Transaction labels have not been implemented.*
- *Transaction counts for links and edges have not been implemented.*
- *Looped transactions not supported.*
- *Some child SVGObjects have id null.*
- *Connection generation produces undesirable results when nodes are placed in close proximity to one another.*
- Transaction arrow shaft does not support diamond line style.

**Possible Improvements:**
- Incorporate the PluginReporter into the plugin.
- The responsibility of setting the dimensions of a SVGObject would be better served in the SVGObject wrapper rather than the SVGGraphBuilder.

*Note: Tests have still not been implemented due to time constraints, and inevitable design changes.*
*Reviewer:  @sol695510*

sol695510 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,342 @@
+     * With respect to the SVGObject it is contained within.
+     * @param posiition
+     */
+    void setSourcePosition(Tuple<Double, Double> position) {
```
Ambiguous method access modifier, specify as public. Also final for parameter.

sol695510 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,342 @@
+     * With respect to the SVGObject it is contained within.
+     * @param position
+     */
+    void setDestinationPosition(Tuple<Double, Double> position) {
```
Ambiguous method access modifier, specify as public. Also final for parameter.

sol695510 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,342 @@
+     * Sets the transformation values of the SVGObject.
+     * @param position
+     */
+    void setTransformation(String transformationData) {
```
Ambiguous method access modifier, specify as public. Also final for parameter.

sol695510 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,342 @@
+     * Sets the color of the SVGObject.
+     * @param color 
+     */
+    public void setFillColor(ConstellationColor color){
```
Final for parameter.

sol695510 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,342 @@
+     * Sets the stoke color of the SVGObject.
+     * @param color 
+     */
+    void setStrokeColor(ConstellationColor color) {
```
Ambiguous method access modifier, specify as public. Also final for parameter.

sol695510 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGObject.java:
```diff
@@ -0,0 +1,342 @@
+     * Sets the stroke style of the SVGObject.
+     * @param style 
+     */
+    void setStrokeStyle(LineStyle style) {
```
Ambiguous method access modifier, specify as public. Also final for parameter.

sol695510 commented:
Overall, the work looks good with what's been achieved in this milestone. Aside from some code changes I've requested, I also have some design suggestions:

### **SVG Export design**
**Current design:**
![Current M7 SVG screenshot](https://github.com/constellation-app/constellation/assets/58677774/da14d128-8cb8-4665-87f7-b75bdec38d58)

**Suggested design:**
![Proposed M7 SVG screenshot](https://github.com/constellation-app/constellation/assets/58677774/031caf07-9c45-4c3c-8bf1-b4591266bb13)

The suggestion to utilise SVG came from the screenshot resolution being dependent on the current screen resolution. Screenshots traditionally did not have this white bordered style of design, and while the title and export time are useful, I suggest they could be added as an optional watermark at the bottom of the image. A checkbox in the Export to SVG window could be used to determine whether to include it or not when the image is exported.

### **Export to SVG window design**
**Current design:**
![Current M7 Export SVG save phase 2-2](https://github.com/constellation-app/constellation/assets/58677774/26221ebc-c51a-4272-a306-fd85476af8ec)

**Suggested design:**
![Current M7 Export SVG save phase 2-2 EDITED](https://github.com/constellation-app/constellation/assets/58677774/37726937-b3de-4783-932c-a779938ec9b1)

The "Graph Title" field seems redundant only because it can get that from the graph itself, unless you wanted a custom title for the SVG exports, up to you, but the current wording seems confusing, so you can always change just that aspect.

Also as mentioned earlier in **SVG Export design** about a watermark for the export details, a checkbox could be included, shown here next to the label "Show Details".

### **Order of operations**
Currently when you perform the action of exporting to an SVG it goes as follows:

![image](https://github.com/constellation-app/constellation/assets/58677774/fba97b9a-0a3f-472a-8751-529f5b115597)

The file dialog opens first then after hitting Save, the Export to SVG window opens. Only the Export to SVG should open, since the window already has the option to browse a file location within it, as is the case for the other Export options windows such as Export > To Shapefile..., Export > To GeoPackage..., Export > To GeoJSON..., etc.

capricornunicorn123 commented:
### Milestone 5 & Milestone 6
**Summary:**
Enables 3D graphs to export with correct position and scale

**Improvements Made:**
- Refactored positional graph calculations to use model-view-perspective matrix calculations. To convert 3D coordinates to a respective projection on a 2D plane.
- Positional Information stored in Vector3f data structure.
- Node background icon color filtering enhanced to better reflect application.
- Plugin progress enabled with details available in plugin reporter.
- Enhanced SVG sizing to ensure size encompasses child elements.
- Basic graph parsing functionality and utility classes moved to Utilities module.

**limitations & Potential Issues:**
- *Code smells relating to OS supportability have been highlighted in the current use of FileReader and InputStreamReader.*
- *SVGParser only permits loading SVG files that utilize single line tags.*
- *The current SVG output is not able to render non Latin characters.*
- *Transaction labels have not been implemented.*
- *Transaction counts for links and edges have not been implemented.*
- *Looped transactions not supported.*
- *Some child SVGObjects have id null.*
- *Connection generation produces undesirable results when nodes are placed in close proximity to one another.*
- *Transaction arrow shaft does not support diamond line style.*
- The larger a graph gets the less proportional the margin becomes, leading to a large amount of empty space. 
- Graph headers and footers reduce in size with increasing graph size. 
- Node scaling is not accurately reflective of the scale factor observed in the application graph view.
- Data entered by uses not parsed meaning illegal characters can be include in the output.

**Possible Improvements:**
- Condense metadata into a smaller header or footer instead of polaroid style border.
- Integrate file location selection into plugin-in parameter window.
- Determine output image view widow dimensions using ModelViewProjection instead of explicit x and y extremities to prepare for exporting from a custom camera angle.

*Note: Tests have still not been implemented due to time constraints, and inevitable design changes.*
*Reviewer:  @Delphinus8821*

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            return this;
+        }
+        
+        public SVGGraphBuilder  withInteraction(PluginInteraction interaction) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         * @param selectedNodesOnly
+         * @return 
+         */
+        public SVGGraphBuilder withNodes(Boolean selectedNodesOnly) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         * @param showConnections
+         * @return 
+         */
+        public SVGGraphBuilder includeConnections(Boolean showConnections) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         * @param showTopLabels
+         * @return 
+         */
+        public SVGGraphBuilder includeTopLabels(Boolean showTopLabels) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         * @param showBottomLabels
+         * @return 
+         */
+        public SVGGraphBuilder includeBottomLabels(Boolean showBottomLabels) {
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         * @param sourcePosition
+         * @param destinationPosition 
+         */
+        private void buildArrowShaft(SVGObject arrowShaft, Vector4f sourcePosition, Vector4f destinationPosition) {
```
finals

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/file/ConstellationInstalledFileLocator.java:
```diff
@@ -48,25 +48,30 @@ public class ConstellationInstalledFileLocator {
     public static File locate(String relativePath, String codeNameBase, ProtectionDomain protectedDomain) {
```
finals

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/ConstellationIcon.java:
```diff
@@ -415,6 +481,51 @@ private ThreeTuple<Integer, Integer, Color> buildCacheKey(final int size, final
+                return os.toByteArray();
+            }
+
+        } catch (IOException ex) {
```
final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,491 @@
+     * @param minimumWidth
+     * @param minimumHeight 
+     */
+    private void setMinimumHeight(float minimumHeight) {
```
final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/IconData.java:
```diff
@@ -29,14 +32,46 @@
+                svg.saturateSVG(ConstellationColor.fromJavaColor(color));
+            }
+            return svg.toSVGData();
+        } catch (Exception ex){
```
final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/ConstellationIcon.java:
```diff
@@ -415,6 +481,51 @@ private ThreeTuple<Integer, Integer, Color> buildCacheKey(final int size, final
+                final ByteArrayOutputStream os = new ByteArrayOutputStream();
+                try {
+                    ImageIO.write(coloredImage, ConstellationIcon.DEFAULT_ICON_FORMAT, os);
+                } catch (IOException ex) {
```
final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/file/ConstellationInstalledFileLocator.java:
```diff
@@ -48,25 +48,30 @@ public class ConstellationInstalledFileLocator {
-            }
+                try {
+                    locatedFile = Paths.get(url.toURI()).toFile();
+                } catch (URISyntaxException ex) {
```
final

Delphinus8821 reviewed (changes requested):
Just first check through for finals, will review functionality later.

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         */
+        private void preBuild(){
+            
+            Camera camera = new Camera(access.getCamera());
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            ymin = -ymax;
+            xmin = -xmax;
+
+            Frustum viewFrustum = new Frustum(fov, aspect, xmin, xmax, ymin, ymax, nearPerspective, farPerspective);
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            Frustum viewFrustum = new Frustum(fov, aspect, xmin, xmax, ymin, ymax, nearPerspective, farPerspective);
+
+            //Get the projection matrix from the view frustum
+            Matrix44f pMatrix = viewFrustum.getProjectionMatrix();
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+                }
+                
+                final SVGObject backgroundContainer = nodeImages.getChild(SVGLayoutConstant.BACKGROUND_IMAGE.getValue());
+                SVGData bgimage = backgroundIcon.buildSVG(color.getJavaColor());
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+                bgimage.setParent(backgroundContainer.toSVGData());
+                
+                final SVGObject foregroundContainer = nodeImages.getChild(SVGLayoutConstant.FOREGROUND_IMAGE.getValue());
+                SVGData fgimage = foregroundIcon.buildSVG();
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            for (int labelPosition = 0; labelPosition < access.getBottomLabelCount(); labelPosition++) {
+                final String labelString = access.getVertexBottomLabelText(vertex, labelPosition);
+                if (labelString != null){
+                    SVGObject text = SVGObject.loadFromTemplate(SVGFileNameConstant.LABEL);
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            final Float constelationGraphY = access.getY(vertex);
+            final Float constelationGraphZ = access.getZ(vertex);
+            
+            Vector3f worldPosition = new Vector3f(constelationGraphX, constelationGraphY, constelationGraphZ);
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+        }
+        
+        private Vector4f getVertexPosition(final Vector3f worldPosition){
+            Vector4f screenPosition = new Vector4f();
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            Vector4f screenPosition = new Vector4f();
+            
+            Graphics3DUtilities.project(worldPosition, modelViewProjectionMatrix, viewPort, screenPosition);
+            Vector4f centerOffSet = new Vector4f(128, 128, 0, 0);
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+         * @return 
+         */
+        private float getVertexScaledRadius(final int vertex) {  
+            Vector4f screenPosition = getVertexPosition(vertex);
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            final Float y = access.getY(vertex) + 1;
+            final Float z = access.getZ(vertex);
+            
+            Vector4f edgePosition = getVertexPosition(new Vector3f(x,y,z));
```
final

Delphinus8821 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java:
```diff
@@ -0,0 +1,732 @@
+            
+            Vector4f edgePosition = getVertexPosition(new Vector3f(x,y,z));
+            
+            int radiusID = VisualConcept.VertexAttribute.NODE_RADIUS.get(graph);
+            float radius = graph.getFloatValue(radiusID, access.getVertexId(vertex));
+            float depthScaleFactor = Math.abs(edgePosition.getX() - screenPosition.getX());
```
all 3 can be final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,491 @@
+    }
+    
+    private Collection<SVGObject> getAllChildren(){
+        ArrayList<SVGObject> children = new ArrayList<>();
```
final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGParser.java:
```diff
@@ -0,0 +1,230 @@
+     */
+    private static String removeNonLatinCharacters(final String text){
+        final StringBuilder builder = new StringBuilder();
+        char[] charArray = text.toCharArray();
```
final

Delphinus8821 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGParser.java:
```diff
@@ -0,0 +1,230 @@
+     */
+    private static Map<String,String> getElementAttributes(final String svgString) {
+        final Map<String,String> extractedAttributes = new LinkedHashMap<>();
+        String regex = "[-:a-zA-Z0-9]*=\"[,\\-/:%#\\s.a-zA-Z0-9]*\"";
+        Pattern svgAttributeAssignmentRegex = Pattern.compile(regex);
+        Matcher svgMatcher = svgAttributeAssignmentRegex.matcher(svgString);
```
all 3 can be final

Delphinus8821 reviewed (changes requested):
More finals

Delphinus8821 commented:
Functionality is working really well. I agree with previous comments made about combining the file selection and parameters panes but looking good overall so far.

capricornunicorn123 commented:
### Milestone 8
 
![Milestone Graphic 8](https://github.com/constellation-app/constellation/assets/139201011/1e2eb715-aa68-4a9e-85a8-f260c52cfd4c)

**Summary:**
Custom Export Perspectives have been enabled along with a simplified exported image format that closer replicates the users view of the graph within the application.

**Improvements Made:**
- Loped transactions now supported.
- Looped transaction image assets enhanced. 
- Enabled export from X, Y, Z and Current View perspectives.
- Refactored SVG layout to simplify graph information.
- Exported SVG dimensions now reflect users view window.
- Nodes and Transactions now render in the correct order. With distant elements behind closer elements. 
- Prevented invisible connections from being exported.
- Export process has been simplified with on pop up window instead of two.
- Enables Connection scaling with depth. 
- Animations added to modify user perspective when exporting .
- Transaction Labels Implemented.

**limitations & Potential Issues:**
- *Code smells relating to OS supportability have been highlighted in the current use of FileReader and InputStreamReader.*
- *SVGParser only permits loading SVG files that utilize single line tags.*
- *The current SVG output is not able to render non Latin characters.*
- *Transaction arrow shaft does not support diamond line style.*
- Labels do not have a background fill as the application does making labels difficult to read in large 3D Graphs.

**Possible Improvements:**
- Enable line styles to calculate the points for a path that would draw the diamond line style.
- Blazes could be included. 
- Image of the x, y, z, axis could be included to give orientation of graph.
- Increase the number of SVG image assets in the application to increase performance.

*Note: Tests have still not been implemented due to time constraints, and inevitable design changes.*
*Reviewer:  @Delphinus8821*

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+    /**
+     * Sets up the Builder control attributes and reference utilities. 
+     */
+    private void preBuild() throws Exception {
```
is there a more specific exception that can be thrown here?

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+        
+        try{
+            preBuild();
+        } catch (Exception e){
```
would enable this catch to also be more specific

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+        access.updateInternally();
+        
+        // Determine the height and width of the users InteractiveGraphView pane
+        int paneHeight = visualManager.getVisualComponent().getHeight();   
+        int paneWidth = visualManager.getVisualComponent().getWidth();
```
both can be final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+
+        // Determine which segment this connection label will occupy
+        final int labelSegment = (connectionIndex % 7) + 1;
+        float segmentRatio = (float) labelSegment / totalSegments;
+        
+        // Calculate the position of the label
+        float distance = getDistance(highPosition, lowPosition); 
+        float offsetDistance = distance * segmentRatio;
+        double angle = this.calculateConnectionAngle(lowPosition, highPosition);
+        Vector4f position = this.offSetPosition(lowPosition, offsetDistance, angle);
```
all 5 can be final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+            
+            //Only add the label if the label value exists.
+            if (labelString != null) {
+                SVGObject svgLabel = SVGTemplateConstant.LABEL.getSVGObject();
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+        Vector4f.add(loopCenterPosition, getVertexPosition(vertexIndex), new Vector4f(getVertexScaledRadius(vertexIndex), -getVertexScaledRadius(vertexIndex), 0, 0));
+
+        // Create the loopedConnection
+        SVGObject svgLoop = SVGTemplateConstant.CONNECTION_LOOP.getSVGObject();
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+    private void buildArrowHead(final SVGObject svgArrowHead, final Vector4f position, final double connectionAngle, final float scaleFactor) {
+        
+        //Set arrow head dimensions
+        float arrowHeadHeight = scaleFactor * 32;
+        float arrowHeadWidth = scaleFactor * 128;
```
both final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+    private void buildLinearArrowShaft(final SVGObject svgArrowShaft, final Vector4f sourcePosition, final Vector4f destinationPosition, final int sourceIndex, final int destinationIndex) {
+        
+        // Get the sclae factor of the source and destination vertices
+        float destinationScaleFactor = this.getDepthScaleFactor(getVertexWorldPosition(destinationIndex));
+        float sourceScaleFactor = this.getDepthScaleFactor(getVertexWorldPosition(sourceIndex));
+
+        // Get the connection angles of the source and destination arrow heads
+        double sourceConnectionAngle = this.calculateConnectionAngle(sourcePosition, destinationPosition);
+        double detinationConnectionAngle = this.calculateConnectionAngle(destinationPosition, sourcePosition);
+
+        // Calculate the four points of the arrow shaft.
+        Vector4f p1 = this.offSetPosition(sourcePosition, 4 * sourceScaleFactor, sourceConnectionAngle + 90);
+        Vector4f p2 = this.offSetPosition(sourcePosition, 4 * sourceScaleFactor, sourceConnectionAngle - 90);
+        Vector4f p3 = this.offSetPosition(destinationPosition, 4 * destinationScaleFactor, detinationConnectionAngle + 90);
+        Vector4f p4 = this.offSetPosition(destinationPosition, 4 * destinationScaleFactor, detinationConnectionAngle - 90);
```
all can be final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+        final Vector4f screenPosition = getScreenPosition(worldPosition);
+
+        // Move the point in the world equivelent of the screens upwards direction by one unit.
+        Vector3f up = camera.lookAtUp;
```
final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+     * @return 
+     */
+    private ConstellationColor getConnectionColor(final int connectionIndex) {
+        final ConstellationColor color;
+        if (access.isConnectionDimmed(connectionIndex)) {
+            color = VisualGraphDefaults.DEFAULT_TRANSACTION_COLOR;
+        } else {
+            color = access.getConnectionColor(connectionIndex);
```
could you convert this all into a ternary statement?

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraphBuilder.java:
```diff
@@ -0,0 +1,948 @@
+     * @return 
+     */
+    private float getDistance(final Vector4f a, final Vector4f b) {
+        float xChange = Math.abs(a.getX()-b.getX());
+        float yChange = Math.abs(a.getY()-b.getY());
```
both final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGObjectConstant.java:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @author capricornunicorn123
+ */
+public enum SVGObjectConstant {
```
add an 's' at the end (Constants)

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGObjectConstant.java:
```diff
@@ -0,0 +1,79 @@
+     * @param parent
+     * @return 
+     */
+    public SVGObject findIn(SVGObject parent) {
```
parameter final

antares1470 commented on CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGObjectConstant.java:
```diff
@@ -0,0 +1,79 @@
+     * Uses depth first search.
+     * @param parent 
+     */
+    public void removeFrom(SVGObject parent) {
```
ditto

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/ConstellationIcon.java:
```diff
@@ -415,6 +481,50 @@ private ThreeTuple<Integer, Integer, Color> buildCacheKey(final int size, final
+                    }
+                }
+                final ByteArrayOutputStream os = new ByteArrayOutputStream();
+                try {
+                    ImageIO.write(coloredImage, ConstellationIcon.DEFAULT_ICON_FORMAT, os);
+                } catch (final IOException ex) {
+                    return original;
+                }
```
given the catch in the outer try is exactly the same, you probably can omit the inner try (if `ImageIO.write` throws an IOException then it will be caught in the outer try)

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/FileIconData.java:
```diff
@@ -24,27 +24,50 @@
+        final File locatedRasterFile = ConstellationInstalledFileLocator.locate(relativePath, codeNameBase, FileIconData.class.getProtectionDomain());
+        this.rasterFile = locatedRasterFile;
+        
+        final File locatedVectorFile = ConstellationInstalledFileLocator.locate(relativePath.replaceAll(".png", ".svg"), codeNameBase, FileIconData.class.getProtectionDomain());
```
can the FileExtensionConstants be used in the replaceAll function

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/FileIconData.java:
```diff
@@ -24,27 +24,50 @@
+    protected InputStream createVectorInputStream() throws IOException {
+        
+        //Not all files have SVG alternatives to only return a stream if the alternative was found.
+        if (vectorFile != null) {
+            return new FileInputStream(vectorFile);
+        } else {
+            return null;
+        }
```
use ternary statement here

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGAttributeConstant.java:
```diff
@@ -0,0 +1,100 @@
+ * 
+ * @author capricornunicorn123
+ */
+public enum SVGAttributeConstant {
```
"Constants"

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGAttributeConstant.java:
```diff
@@ -0,0 +1,100 @@
+     * @return 
+     */
+    public static Map<String, String> initialiseBasicAttributes() {
+        Map<String, String> map = new LinkedHashMap<>();
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGAttributeConstant.java:
```diff
@@ -0,0 +1,100 @@
+     * Created for the SVGData constructor to ensure that significant attributes such as id, x and y are forced to the front of the svg tag. 
+     * @return 
+     */
+    public static Map<String, String> initialiseBasicAttributes() {
```
just looking at the usage of this function, I'm not sure if it is entirely needed as when you retrieve values from a map, it will return null anyway if the key doesn't exist (and if you populate value with other values, that doesn't need an explicit initialise first)

capricornunicorn123 replied:
The LinkedHashMap that stores attributes ensures that element addition order is maintained with the intention of allowing the resulting svg string to have a predictable order of attributes. This has the side effect of not allowing developers to add attributes willy-nilly as the order of addition will effect the order of output - example of ad hoc element addition on output.

- <svg x="4" viewBox="0, 0, 32, 32" width="8" fill="$fff000" id="myID" y="7" />

As adding attributes in an ad hoc manner is convenient for development it was useful to have important attributes such as identifying, positional and color information at the front of the output attributes and in a predictable order. 

This map initialization ensures that the order of these common element are predictable. When converting to a string, the SVGData object "cleans" the attributes and removes and key value pairs from the map that have a value of null.

Removing this method will not add or detract from the visual output of the plugin. Is it worth keeping this to aid in future development or sould it be removed?

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+    private SVGData parent;
+
+    public SVGData(final String type, final SVGData parent, final Map<String, String> attributes) {
+        if (type == null){
+            throw new NullPointerException("SVGData elements cannot have a type Null");
+        }
```
Throwing null pointers isn't usually a good practice. Might be best in this case to just surround the remaining code in a `(type != null)` and leave at that
Alternatively, perhaps the exception you want to use here is `IllegalArgumentException`

capricornunicorn123 replied:
I've had a think about what I am trying to achieve here and as an alternate design I will create an enum with a set of valid svg types and enforce that the svg type must be one of these tags. That way it assist future developers and avoids throwing errors.

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+    public SVGData getChild(final String idValue) {
+        SVGData child = this.children.get(idValue);
+        if (child == null) {
+            for (SVGData childIndex : this.getAllChildren()) {
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+        if (child == null) {
+            for (SVGData childIndex : this.getAllChildren()) {
+                child = childIndex.getChild(idValue);
+                if (child != null) {
+                    return child;
```
just noting the return here is the same as it is out of the loop. Perhaps just adding a break here is what is needed? (probably not essential but a thought)

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+        
+        //If the child wasnt removed, recursively itterate through levels of children untill the id is found.
+        if (child == null) {
+            for (SVGData childIndex : this.getAllChildren()) {
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+        children.keySet().forEach(key -> allChildren.add(this.children.get(key)));
+        if (!allChildren.isEmpty() && allChildren.get(0).attributes.keySet().contains(SVGAttributeConstant.CUSTOM_SORT_ORDER.getName())) {
+            Collections.sort(allChildren, (SVGData first, SVGData second) -> {
+                String firstValue = first.getAttributeValue(SVGAttributeConstant.CUSTOM_SORT_ORDER.getName());
+                String secondValue = second.getAttributeValue(SVGAttributeConstant.CUSTOM_SORT_ORDER.getName());
```
both final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+        
+        if (parent != null) {            
+            if (parent.getChild(this.getAttributeValue(idAttributeName)) != null){
+                throw new ArrayIndexOutOfBoundsException(String.format("Parent SVGData %s already has a child of id %s", 
```
similar case with Null pointer, not sure if array out of the bounds is the correct exception here. Perhaps `IllegalStateException`?

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,346 @@
+        try {
+            templateSVG = SVGParser.parse(inputStream);
+        } catch (final IOException ex) {
+            LOGGER.log(Level.INFO, ex.getLocalizedMessage());
```
make log level SEVERE

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+    }
+    
+    public SVGObject getParent() {
+        SVGData parent = this.svgDataReference.getParent();
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+    
+    public SVGObject getParent() {
+        SVGData parent = this.svgDataReference.getParent();
+        if (parent == null){
+            return null;
+        } else {
+            return new SVGObject(this.svgDataReference.getParent());
```
can make ternary

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+     */
+    public final SVGObject getChild(final String id) {
+        final SVGData child = this.svgDataReference.getChild(id);
+        if (child != null) {
+            return new SVGObject(child); 
+        } else {
+            return null;
```
same here

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+    }
+    
+    private Collection<SVGObject> getAllChildren() {
+        final ArrayList<SVGObject> children = new ArrayList<>();
```
can you declare just as plain List?

antares1470 replied:
just noting I don't think this one was done

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+     * @return 
+     */
+    private Float getAttributeFloat(final SVGAttributeConstant attributeKey) {
+        String attribute = getAttributeString(attributeKey);
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+        String attribute = getAttributeString(attributeKey);
+        try {
+            return Float.parseFloat(attribute);
+        } catch (final NumberFormatException | NullPointerException e) {
```
add a null check rather than catching null pointer

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,471 @@
+    public static SVGObject loadFromInputStream(InputStream is) {
+        try{
+            return new SVGObject(SVGParser.parse(is));
+        } catch (IOException e) {
```
final

antares1470 replied:
I think this one was missed too

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGParser.java:
```diff
@@ -0,0 +1,219 @@
+
+                // Create a new SVGData with the current SVGData as the parent 
+                if (openTag) {
+                    SVGData newObject = new SVGData(
```
final

antares1470 reviewed (changes requested):
Just looking at your unit test classes, one thing to note is that Constellation is currently using NGTest rather than JUnit

If you could add a whatsnew entry for this, that would be great

OrionsGuardian commented:
A few small things I noticed while testing :

![image](https://github.com/constellation-app/constellation/assets/104604488/c5b29cdd-0f9a-42a0-afc5-d027d7037f21)

capricornunicorn123 commented:
> A few small things I noticed while testing :

I agree with your comments but am cautious that some of these changes may be outside of the scope of this ticket. 

**FileLocation Save Button**
This button is defined in the ExportToSVGAction using a FileParameterKind enum. This enum has 3 values SAVE, OPEN & OPEN_MULTIPLE. The enum also helps to control the functionality of the file selection popup so simply adding a new kind to change the button text may not be a robust solution. This in conjunction with the file chooser problems in issue #1825 may warrant adding this problem to that issue and reviewing all FileChooser windows / FileParameterType in the application.

**Selected Elements**
In previous discussions it was concluded that transactions are visually insignificant without their corresponding nodes. Based on this, we restricted exporting to be based on nodes and transactions brought along accordingly. This is why we only allow selected nodes or all nodes to be exported. We can add functionality where-by selected transactions are exported and corresponding nodes are brought along, but this may not be useful if those cases do not exist with normal use. Are there cases where you may query a graph and select transactions based on the query and then want data exported based on these selected transactions?

**PluginParameter OK Button**
The Pop-Up window allowing for parameters to be customized is generated by the DefaultPluginEnvironment through the DefaultPluginInteraction.prompt(). this is achieved by using the PluginExecution.Interacively(true) method at the ExportToSVGAction level. Basically the labelling of this button is not available to customisation at the individual plugin level. If work were to be put in to customising this button, it would have a potential impact on all other plugins that use their plugin parameters interactively and may be worth investigating what other button names are worth including. I’d be inclined to capture this in a separate ticket.

antares1470 commented on CoreImportExportPlugins/test/unit/src/au/gov/asd/tac/constellation/plugins/importexport/svg/resources/SVGTemplateConstantsNGTest.java:
```diff
@@ -0,0 +1,68 @@
+ */
+package au.gov.asd.tac.constellation.plugins.importexport.svg.resources;
+
+import static org.testng.Assert.*;
```
replace with specific uses of library

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/graphics/Vector4f.java:
```diff
@@ -134,6 +134,16 @@ public static void subtract(final Vector4f result, final Vector4f a, final Vecto
+    public static void reflect(final Vector4f result, final Vector4f referencePoint, final Vector4f reflectionPoint) {
+        
+        // Determine the distance between the reference point and reflection point
+        Vector4f distance = new Vector4f();
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/visual/AxisConstants.java:
```diff
@@ -0,0 +1,60 @@
+    
+    public static AxisConstants getReference(final String name){
+        if (name != null){
+            for (AxisConstants value : AxisConstants.values()){
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/visual/AxisConstants.java:
```diff
@@ -0,0 +1,60 @@
+    public static AxisConstants getReference(final String name){
+        if (name != null){
+            for (AxisConstants value : AxisConstants.values()){
+                if (value != null){
+                    if (name.equals(value.toString())){
```
these 2 ifs can be merged

antares1470 commented on CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/visual/AxisConstantsNGTest.java:
```diff
@@ -0,0 +1,72 @@
+ */
+package au.gov.asd.tac.constellation.utilities.visual;
+
+import static org.testng.Assert.*;
```
replace with specific uses of library

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/icon/IconData.java:
```diff
@@ -29,14 +32,46 @@
+                svg.saturateSVG(ConstellationColor.fromJavaColor(color));
+            }
+            return svg.toSVGData();
+        } catch (final Exception ex) {
```
replace with more specific exception catch

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,355 @@
+        final String linePrefix = SeparatorConstants.NEWLINE + (prefix == null ? "" : prefix);
+        final Set<String> keys = attributes.keySet();
+        keys.forEach(key -> {
+            if (SVGAttributeConstants.NAME_SPACE.getName().equals(key) && this.parent != null) {
+                //do nothing here
+            } else {
+                attributeBuilder.append(String.format(" %s=\"%s\"", key, attributes.get(key)));
```
would it be easier to read if you were to flip this which would remove the need for an else clause?

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,355 @@
+        final StringBuilder attributeBuilder = new StringBuilder();
+        final Set<String> keys = attributes.keySet();
+        keys.forEach(key -> {
+            if (SVGAttributeConstants.NAME_SPACE.getName().equals(key) && this.parent != null) {
+                //do nothing here
+            } else {
+                attributeBuilder.append(String.format(" %s=\"%s\"", key, attributes.get(key)));
```
similar

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGData.java:
```diff
@@ -0,0 +1,355 @@
+     */
+    private String elementChildrenToSVG (final String prefix) {
+        final StringBuilder childSVGString = new StringBuilder();
+        String childPrefix = SeparatorConstants.TAB;
+        if (prefix != null) {
+            childPrefix += prefix;
```
use a StringBuilder here

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,517 @@
+     * @param attributeValue 
+     */
+    private void setAttribute(final SVGAttributeConstants attributeKey, final int attributeValue) {
+        svgDataReference.setAttribute(attributeKey, String.format("%s", attributeValue));
```
instead of using a string formatter here, could use `String.valueOf(attributeValue)`

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,517 @@
+     * @param attributeValue 
+     */
+    private void setAttribute(final SVGAttributeConstants attributeKey, final ConstellationColor attributeValue) {
+        svgDataReference.setAttribute(attributeKey, String.format("%s", attributeValue.getHtmlColor()));
```
`getHtmlColor` already returns a string so need to use a formatter here

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/svg/SVGObject.java:
```diff
@@ -0,0 +1,517 @@
+     * @param points 
+     */
+    public void setPoints(final Vector4f... points) {
+        StringBuilder sb = new StringBuilder();
```
final

antares1470 commented on CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/icon/ByteIconDataNGTest.java:
```diff
@@ -0,0 +1,77 @@
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import static org.testng.Assert.*;
```
replace with specific imports

antares1470 commented on CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/svg/SVGTypeConstantsNGTest.java:
```diff
@@ -0,0 +1,64 @@
+ */
+package au.gov.asd.tac.constellation.utilities.svg;
+
+import static org.testng.Assert.*;
```
same here
